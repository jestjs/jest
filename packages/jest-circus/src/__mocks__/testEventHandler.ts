/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {Circus} from '@jest/types';

const testEventHandler: Circus.EventHandler = (event, state) => {
  switch (event.name) {
    case 'start_describe_definition':
    case 'finish_describe_definition': {
      console.log(`${event.name}:`, event.blockName);
      break;
    }
    case 'run_describe_start':
    case 'run_describe_finish': {
      console.log(`${event.name}:`, event.describeBlock.name);
      break;
    }
    case 'test_start':
    case 'test_started':
    case 'test_retry':
    case 'test_done': {
      console.log(`${event.name}:`, event.test.name);
      break;
    }

    case 'add_test': {
      console.log(`${event.name}:`, event.testName);
      break;
    }

    case 'test_fn_start':
    case 'test_fn_success':
    case 'test_fn_failure': {
      console.log(`${event.name}:`, event.test.name);
      break;
    }

    case 'add_hook': {
      console.log(`${event.name}:`, event.hookType);
      break;
    }

    case 'hook_start':
    case 'hook_success':
    case 'hook_failure': {
      console.log(`${event.name}:`, event.hook.type);
      break;
    }

    default:
      console.log(event.name);
  }

  if (event.name === 'run_finish') {
    console.log('');
    console.log(`unhandledErrors: ${String(state.unhandledErrors.length)}`);
  }
};

export default testEventHandler;
