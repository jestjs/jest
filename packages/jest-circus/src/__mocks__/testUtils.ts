/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {sync as spawnSync} from 'execa';
import * as fs from 'graceful-fs';
import tempy = require('tempy');

const CIRCUS_PATH = require.resolve('../').replaceAll('\\', '\\\\');
const CIRCUS_RUN_PATH = require.resolve('../run').replaceAll('\\', '\\\\');
const CIRCUS_STATE_PATH = require.resolve('../state').replaceAll('\\', '\\\\');
const TEST_EVENT_HANDLER_PATH = require
  .resolve('./testEventHandler')
  .replaceAll('\\', '\\\\');
const BABEL_REGISTER_PATH = require
  .resolve('@babel/register')
  .replaceAll('\\', '\\\\');

export const runTest = (
  source: string,
  opts?: {seed?: number; randomize?: boolean},
) => {
  const tmpFilename = tempy.file();

  const content = `
    require('${BABEL_REGISTER_PATH}')({extensions: [".js", ".ts"]});
    const circus = require('${CIRCUS_PATH}');
    global.test = circus.test;
    global.describe = circus.describe;
    global.beforeEach = circus.beforeEach;
    global.afterEach = circus.afterEach;
    global.beforeAll = circus.beforeAll;
    global.afterAll = circus.afterAll;

    const testEventHandler = require('${TEST_EVENT_HANDLER_PATH}').default;
    const {addEventHandler, removeEventHandler, getState} = require('${CIRCUS_STATE_PATH}');
    getState().randomize = ${opts?.randomize};
    getState().seed = ${opts?.seed ?? 0};
    addEventHandler(testEventHandler);

    ${source};

    const run = require('${CIRCUS_RUN_PATH}').default;

    run();
  `;

  fs.writeFileSync(tmpFilename, content);
  const result = spawnSync('node', [tmpFilename], {
    cwd: process.cwd(),
  });

  if (result.exitCode !== 0) {
    const message = `
      STDOUT: ${result.stdout && result.stdout.toString()}
      STDERR: ${result.stderr && result.stderr.toString()}
      STATUS: ${result.exitCode}
    `;
    throw new Error(message);
  }

  fs.rmSync(tmpFilename, {force: true});

  if (result.stderr) {
    throw new Error(
      `
      Unexpected stderr:
      ${result.stderr}
    `,
    );
  }
  return result;
};
