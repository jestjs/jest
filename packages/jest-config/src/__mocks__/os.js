/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

const os = jest.createMockFromModule('os');

let cpus;
function __setCpus(newCpus) {
  cpus = newCpus;
}

os.__setCpus = __setCpus;
os.cpus = jest.fn(() => cpus);
os.availableParallelism = jest.fn(() => cpus?.length ?? 0);

module.exports = os;
