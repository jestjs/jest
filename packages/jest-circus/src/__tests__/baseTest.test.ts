/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {runTest} from '../__mocks__/testUtils';

test('simple test', () => {
  const {stdout} = runTest(`
    describe('describe', () => {
      beforeEach(() => {});
      afterEach(() => {});
      test('one', () => {});
      test('two', () => {});
    })
  `);

  expect(stdout).toMatchSnapshot();
});

test('function descriptors', () => {
  const {stdout} = runTest(`
    describe(function describer() {}, () => {
      test(class One {}, () => {});
    })
  `);

  expect(stdout).toMatchSnapshot();
});

test('failures', () => {
  const {stdout} = runTest(`
    describe('describe', () => {
      beforeEach(() => {});
      afterEach(() => { throw new Error('banana')});
      test('one', () => { throw new Error('kentucky')});
      test('two', () => {});
    })
  `);

  expect(stdout).toMatchSnapshot();
});

test('concurrent', () => {
  const {stdout} = runTest(`
    describe('describe', () => {
      beforeEach(() => {});
      afterEach(() => { throw new Error('banana')});
      test.concurrent('one', () => { 
        console.log('hello one');
        throw new Error('kentucky')
      });
      test.concurrent('two', () => {
        console.log('hello two');
      });
      test.concurrent('three', async () => { 
        console.log('hello three');
        await Promise.resolve(); 
      });
    })
  `);

  expect(stdout).toMatchSnapshot();
});

test('concurrent.each', () => {
  const {stdout} = runTest(`
    describe('describe', () => {
      beforeEach(() => {});
      afterEach(() => { throw new Error('banana')});
      test.concurrent.each([
        ['one'],
        ['two'],
        ['three'],
      ])('%s', async (name) => {
        console.log('hello %s', name);
        await Promise.resolve(); 
      });
    })
  `);

  expect(stdout).toMatchSnapshot();
});
