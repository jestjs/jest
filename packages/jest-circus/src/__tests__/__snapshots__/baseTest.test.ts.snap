// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`concurrent 1`] = `
"start_describe_definition: describe
add_hook: beforeEach
add_hook: afterEach
add_test: one
add_test: two
add_test: three
finish_describe_definition: describe
run_start
run_describe_start: ROOT_DESCRIBE_BLOCK
run_describe_start: describe
test_start: one
test_start: two
test_start: three
test_started: one
test_started: two
test_started: three
test_fn_start: one
test_fn_start: two
test_fn_start: three
hello one
hello two
hello three
test_fn_failure: one
test_fn_success: two
test_fn_success: three
test_done: one
test_done: two
test_done: three
run_describe_finish: describe
run_describe_finish: ROOT_DESCRIBE_BLOCK
run_finish

unhandledErrors: 0"
`;

exports[`concurrent.each 1`] = `
"start_describe_definition: describe
add_hook: beforeEach
add_hook: afterEach
add_test: one
add_test: two
add_test: three
finish_describe_definition: describe
run_start
run_describe_start: ROOT_DESCRIBE_BLOCK
run_describe_start: describe
test_start: one
test_start: two
test_start: three
test_started: one
test_started: two
test_started: three
test_fn_start: one
test_fn_start: two
test_fn_start: three
hello one
hello two
hello three
test_fn_success: one
test_fn_success: two
test_fn_success: three
test_done: one
test_done: two
test_done: three
run_describe_finish: describe
run_describe_finish: ROOT_DESCRIBE_BLOCK
run_finish

unhandledErrors: 0"
`;

exports[`failures 1`] = `
"start_describe_definition: describe
add_hook: beforeEach
add_hook: afterEach
add_test: one
add_test: two
finish_describe_definition: describe
run_start
run_describe_start: ROOT_DESCRIBE_BLOCK
run_describe_start: describe
test_start: one
test_started: one
hook_start: beforeEach
hook_success: beforeEach
test_fn_start: one
test_fn_failure: one
hook_start: afterEach
hook_failure: afterEach
test_done: one
test_start: two
test_started: two
hook_start: beforeEach
hook_success: beforeEach
test_fn_start: two
test_fn_success: two
hook_start: afterEach
hook_failure: afterEach
test_done: two
run_describe_finish: describe
run_describe_finish: ROOT_DESCRIBE_BLOCK
run_finish

unhandledErrors: 0"
`;

exports[`function descriptors 1`] = `
"start_describe_definition: describer
add_test: One
finish_describe_definition: describer
run_start
run_describe_start: ROOT_DESCRIBE_BLOCK
run_describe_start: describer
test_start: One
test_started: One
test_fn_start: One
test_fn_success: One
test_done: One
run_describe_finish: describer
run_describe_finish: ROOT_DESCRIBE_BLOCK
run_finish

unhandledErrors: 0"
`;

exports[`simple test 1`] = `
"start_describe_definition: describe
add_hook: beforeEach
add_hook: afterEach
add_test: one
add_test: two
finish_describe_definition: describe
run_start
run_describe_start: ROOT_DESCRIBE_BLOCK
run_describe_start: describe
test_start: one
test_started: one
hook_start: beforeEach
hook_success: beforeEach
test_fn_start: one
test_fn_success: one
hook_start: afterEach
hook_success: afterEach
test_done: one
test_start: two
test_started: two
hook_start: beforeEach
hook_success: beforeEach
test_fn_start: two
test_fn_success: two
hook_start: afterEach
hook_success: afterEach
test_done: two
run_describe_finish: describe
run_describe_finish: ROOT_DESCRIBE_BLOCK
run_finish

unhandledErrors: 0"
`;
